### PR TITLE
New resource `azurerm_api_management_api_release`

### PR DESCRIPTION
**Pull request by yupwei68**
Fix https://github.com/terraform-providers/terraform-provider-azurerm/issues/12428

--- PASS: TestAccApiManagementApiRelease_basic (285.27s)
-- PASS: TestAccApiManagementApiRelease_requiresImport (302.57s)
--- PASS: TestAccApiManagementApiRelease_complete (286.16s)
--- PASS: TestAccApiManagementApiRelease_update (513.02s)

### CODE REVIEW

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] apimanagement %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving %q: %+v", id, err)
+	}
+	d.Set("name", id.ReleaseName)
+	if props := resp.APIReleaseContractProperties; props != nil {
+		d.Set("api_id", props.APIID)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:122]: Can we construct a `parse.ApiId{}` and use it to set the `api_id` attribute?

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	id, err := parse.ApiReleaseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.ServiceName, id.ApiName, id.ReleaseName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] apimanagement %q does not exist - removing from state", d.Id())
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:114]:
```suggestion
			log.Printf("[INFO] apimanagement %s does not exist - removing from state", d.Id())
```

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+	parameters := apimanagement.APIReleaseContract{
+		APIReleaseContractProperties: &apimanagement.APIReleaseContractProperties{
+			APIID: utils.String(d.Get("api_id").(string)),
+			Notes: utils.String(d.Get("notes").(string)),
+		},
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, apiId.ResourceGroup, apiId.ServiceName, apiId.Name, name, parameters, ""); err != nil {
+		return fmt.Errorf("creating/ updating %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:94]:
```suggestion
		return fmt.Errorf("creating/ updating %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	if err != nil {
+		return err
+	}
+	id := parse.NewApiReleaseID(subscriptionId, apiId.ResourceGroup, apiId.ServiceName, apiId.Name, name)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, apiId.ResourceGroup, apiId.ServiceName, apiId.Name, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for existing %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:78]:
```suggestion
				return fmt.Errorf("checking for existing %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.ServiceName, id.ApiName, id.ReleaseName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] apimanagement %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:118]:
```suggestion
		return fmt.Errorf("retrieving %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.ApiReleaseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	if _, err := client.Delete(ctx, id.ResourceGroup, id.ServiceName, id.ApiName, id.ReleaseName, "*"); err != nil {
+		return fmt.Errorf("deleting %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:139]:
```suggestion
		return fmt.Errorf("deleting %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_api_release.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	parameters := apimanagement.APIReleaseContract{
+		APIReleaseContractProperties: &apimanagement.APIReleaseContractProperties{
+			APIID: utils.String(d.Get("api_id").(string)),
+			Notes: utils.String(d.Get("notes").(string)),
+		},
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, apiId.ResourceGroup, apiId.ServiceName, apiId.Name, name, parameters, ""); err != nil {
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_api_release.go:93]: The method docstring for `APIReleaseClient{}.CreateOrUpdate()` says that `ifMatch` is required when updating a release; should it perhaps be `"*"` like it is for the `Delete()` method, or something else?

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @yupwei68, thanks for this PR. I made some small suggestions below, if you can look at these and check the final argument for the CreateOrUpdate method, this should be good to merge assuming tests are passing. Thanks!

**Comment by yupwei68** [comment]
@manicminer Thanks. Please continue reviewing.

**Review by manicminer — APPROVED** [review]
Thanks @yupwei68, this LGTM 👍 

Screenshot 2021-07-15 at 11 56 39

**Pull request merged**